File: inhouse/cpg.jl
# 786

function is_cpg(s)
   if s == 'C' 1 
   elseif s == 'G' 2
   else 0
   end
end

m, M = 99999, 0
function cpg(s)
   i, cnt = 1, 0
   while i <= length(s)
      cc, l = 0, 0
      while i <= length(s) 
         c = is_cpg(s[i])
         if c == 0 break end
         cc |= c
         i += 1
         l += 1
      end
      if cc == 3
         cnt += 1
         global m = min(m,l)
         global M = max(M,l)
      end
      i += 1
   end
   cnt
end

function main(arg)
   f = open(arg)
   cnt = 0
   for l in eachline(f)
      cnt += cpg(l)
   end
   println("$cnt $m $M")
end

main(ARGS[1])

File: paper/cpg/cpg.cc
// 786
// benchmarks:
// - count number of CpG islands (e.g. all CG-only subseqs)

#include <iostream>
#include <fstream>
#include <vector>
using namespace std;

char is_cpg(char s) {
   if (s == 'C') return 1; 
   if (s == 'G') return 2;
   return 0;
}

int m=99999, M=0;
long long cpg_count(string s) {
   long long count = 0;
   int i = 0;
   char c;
   while (i < s.size()) {
      char cc = 0; int l = 0;
      while (i < s.size() && (c = is_cpg(s[i]))) {
         cc |= c;
         i++; l++;
      }
      if (cc == 3) {
         count ++;
         m=min(m, l);
         M=max(M, l);
      }
      i++;
   }
   return count;
}

int main(int argc, char **argv) {
   ios_base::sync_with_stdio(false);
	 cin.tie(nullptr);

   ifstream fin(argv[1]);
   string s;
   long long res = 0;
   while (getline(fin, s)) {
      res += cpg_count(s);
   }
   cout << res << ' ' << m << ' ' << M << endl;

   return 0;
}
